Accessor Functionality for AnnData

srivarra opened the pull request:
<!-- Please:
1. Fill in the following check boxes
2. Make sure checks pass (Ignore “Triage” ones)
-->

- [x] Closes #1869
- [x] Tests added
- [x] Release note added (or unnecessary)


Added Accessor / Extension functionality based off the work in `Polars`. Added tests as well.

Added `_accessors` to `anndata._core.anndata.AnnData` to house registered accessor names to avoid name conflicts.

You can create an accessor by decorating a class with methods with the `register_anndata_namespace` function.

Here's an example of how to use it:

```python
import anndata as ad

@ad.register_anndata_namespace("my_accessor")
class Greetings:
	def __init__(self, adata: ad.AnnData):
		self._adata = adata
	
	def greet():
		return "hi"

# and to use

adata.my_accessor.greet()
```
Not sure if this needs a dedicated tutorial notebook, or specific documentation page beyond the docstring.

ilan-gold commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,139 @@
+    Taken from:
+    https://github.com/pandas-dev/pandas/blob/ab89c53f48df67709a533b6a95ce3d911871a0a8/pandas/util/_exceptions.py#L30-L51
+    """
+    import anndata as ad
```
Why an internal-to-the-function import?

srivarra replied:
No reason in particular, it was just what Polars used. As long as the function gets the location of the `anndata/__init__.py` file it'll work. I'll move it outside.

srivarra replied:
It is changed, I moved it outside the function in one of the later pushes.

ilan-gold commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,139 @@
+
+# Currently, there are no reserved namespaces internally, but if there ever are,
+# this will not allow them to be overridden.
+_reserved_namespaces: set[str] = set.union(*(cls._accessors for cls in (AnnData,)))
```
```suggestion
_reserved_namespaces: set[str] = set(AnnData._accessors)
```
why not?

srivarra replied:
That should work as well, in addition I forgot to add all the attributes of AnnData so those don't get overwritten as you've stated [here][comment]. Will add that like so:

```python
_reserved_namespaces: set[str] = set(AnnData._accessors) | set(dir(ad.AnnData))
```

[comment]:
https://github.com/scverse/anndata/pull/1870#discussion_r1967424266

ilan-gold commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,139 @@
+            msg = f"cannot override reserved namespace {name!r}"
+            raise AttributeError(msg)
+
+        elif hasattr(cls, name):
```
I would think we want to disallow this behavior in some form.  I wouldn't want someone overriding `X`.  Maybe I'm misunderstanding, but I also don't see a test for this

srivarra replied:
Oh yeah, just tested it out and it absolutely overrides X. What attributes should be protected, should I just do all of them with `dir(AnnData)`?

ilan-gold replied:
Unless there's a reason to allow overriding already existing attributes, why not just throw an attribute error?

ilan-gold replied:
I see this was resolved but it's still just a warning?

srivarra replied:
There is a check right above that for reserved namespaces (i.e. `dir(AnnData)`. 

```python
if name in _reserved_namespaces:
	msg = f"cannot override reserved attribute {name!r}"
  	raise AttributeError(msg)
```

ilan-gold replied:
Ah ok, maybe I was thinking of an older version of the code.  Great!

ilan-gold commented on tests/test_extensions.py:
```diff
@@ -0,0 +1,134 @@
+        UserWarning, match="Overriding existing custom namespace 'dummy'"
+    ):
+
+        @extensions.register_anndata_namespace("dummy")
```
I would only use the public api here and elsewhere
```suggestion
        @ad.register_anndata_namespace("dummy")
```

ilan-gold commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,139 @@
+def _create_namespace(name: str, cls: type[AnnData]) -> Callable[[type[NS]], type[NS]]:
+    """Register custom namespace against the underlying AnnData class."""
+
+    def namespace(ns_class: type[NS]) -> type[NS]:
```
It seems like we could put a stronger guarantee on `ns_class` (and the other generics that use this) to enforce the correct `__init__` method using a `Protocol`, no?

srivarra replied:
Yes this should be possible

ilan-gold reviewed (commented):
Very cool! Some preliminary comments/changes

srivarra commented:
@ilan-gold

Made the requested changes:

- Added `ExtensionNamespace` Protocol:
	- Provides a standardized interface for static type checking with mypy (can't really enforce implementation details when a user builds an accessor with a Protocol)
	- Defines the expected method signatures for custom accessors
- Implemented Runtime Validation via `_check_namespace_signature`:
	- Verifies at registration time that custom accessors have the correct method signatures
 	- Ensures `__init__` properly accepts self and an `adata` parameter of type `anndata.AnnData`
	- Prevents namespace collisions with existing AnnData attributes (e.g., `X`, `obs`, `var_names`)
	- Raises error messages when validation fails

Added tests for (probably) all validation scenarios (missing parameters, wrong parameter names, incorrect type annotations) and tests for namespace collision detection.


A couple of issues I'm having:

I'm not too sure what's going on with the `readthedocs` build, since I'm unable to replicate it locally, instead I have a different error:

<details>
<summary>Here's my error log for `hatch run docs:build`</summary>

```sh
Running Sphinx v8.2.1
loading translations [en]... done
matplotlib is not installed, social cards will not be generated
myst v4.0.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions={'html_image'}, disable_syntax=[], all_links_external=False, links_external_new_tab=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, fence_as_directive=set(), number_code_blocks=[], title_to_header=False, heading_anchors=3, heading_slug_func=None, html_meta={}, footnote_sort=True, footnote_transition=True, words_per_minute=200, substitutions={}, linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', enable_checkboxes=False, suppress_warnings=[], highlight_code_blocks=True)
loading intersphinx inventory 'awkward' from https://awkward-array.org/doc/stable/objects.inv ...
loading intersphinx inventory 'cupy' from https://docs.cupy.dev/en/stable/objects.inv ...
loading intersphinx inventory 'dask' from https://docs.dask.org/en/stable/objects.inv ...
loading intersphinx inventory 'h5py' from https://docs.h5py.org/en/latest/objects.inv ...
loading intersphinx inventory 'hdf5plugin' from https://hdf5plugin.readthedocs.io/en/latest/objects.inv ...
loading intersphinx inventory 'loompy' from https://linnarssonlab.org/loompy/objects.inv ...
loading intersphinx inventory 'numpy' from https://numpy.org/doc/stable/objects.inv ...
loading intersphinx inventory 'pandas' from https://pandas.pydata.org/pandas-docs/stable/objects.inv ...
loading intersphinx inventory 'python' from https://docs.python.org/3/objects.inv ...
loading intersphinx inventory 'scipy' from https://docs.scipy.org/doc/scipy/objects.inv ...
loading intersphinx inventory 'sklearn' from https://scikit-learn.org/stable/objects.inv ...
loading intersphinx inventory 'xarray' from https://docs.xarray.dev/en/stable/objects.inv ...
loading intersphinx inventory 'zarr' from https://zarr.readthedocs.io/en/v2.18.4/objects.inv ...
[autosummary] generating autosummary for: _key_contributors.rst, api.md, benchmark-read-write.ipynb, benchmarks.md, concatenation.rst, contributing.md, fileformat-prose.md, generated/anndata.AnnData.T.rst, generated/anndata.AnnData.X.rst, generated/anndata.AnnData.chunk_X.rst, ..., generated/anndata.settings.override.rst, generated/anndata.settings.rst, generated/anndata.typing.AxisStorable.rst, generated/anndata.typing.Index.rst, generated/anndata.typing.RWAble.rst, index.md, interoperability.md, references.rst, release-notes/index.md, tutorials/index.md
Writing evaluated template result to /Users/srivarra/Dev/Projects/Open Source Contributions/scverse/anndata/docs/_build/html/_static/nbsphinx-code-cells.css
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 114 source files that are out of date
updating environment: [new config] 114 added, 0 changed, 0 removed
reading sources... [  1%] _key_contributors
reading sources... [  2%] api
reading sources... [  3%] benchmark-read-write
reading sources... [  4%] benchmarks
reading sources... [  4%] concatenation
reading sources... [  5%] contributing
reading sources... [  6%] fileformat-prose
reading sources... [  7%] generated/anndata.AnnData
reading sources... [  8%] generated/anndata.AnnData.T
reading sources... [  9%] generated/anndata.AnnData.X
reading sources... [ 10%] generated/anndata.AnnData.chunk_X
reading sources... [ 11%] generated/anndata.AnnData.chunked_X
reading sources... [ 11%] generated/anndata.AnnData.concatenate
reading sources... [ 12%] generated/anndata.AnnData.copy
reading sources... [ 13%] generated/anndata.AnnData.filename
reading sources... [ 14%] generated/anndata.AnnData.is_view
reading sources... [ 15%] generated/anndata.AnnData.isbacked
reading sources... [ 16%] generated/anndata.AnnData.layers
reading sources... [ 17%] generated/anndata.AnnData.n_obs
reading sources... [ 18%] generated/anndata.AnnData.n_vars
reading sources... [ 18%] generated/anndata.AnnData.obs
reading sources... [ 19%] generated/anndata.AnnData.obs_keys
reading sources... [ 20%] generated/anndata.AnnData.obs_names
reading sources... [ 21%] generated/anndata.AnnData.obs_names_make_unique
reading sources... [ 22%] generated/anndata.AnnData.obs_vector
reading sources... [ 23%] generated/anndata.AnnData.obsm
reading sources... [ 24%] generated/anndata.AnnData.obsm_keys
reading sources... [ 25%] generated/anndata.AnnData.obsp
reading sources... [ 25%] generated/anndata.AnnData.raw
reading sources... [ 26%] generated/anndata.AnnData.rename_categories
reading sources... [ 27%] generated/anndata.AnnData.shape
reading sources... [ 28%] generated/anndata.AnnData.strings_to_categoricals
reading sources... [ 29%] generated/anndata.AnnData.to_df
reading sources... [ 30%] generated/anndata.AnnData.to_memory
reading sources... [ 31%] generated/anndata.AnnData.transpose
reading sources... [ 32%] generated/anndata.AnnData.uns
reading sources... [ 32%] generated/anndata.AnnData.uns_keys
reading sources... [ 33%] generated/anndata.AnnData.var
reading sources... [ 34%] generated/anndata.AnnData.var_keys
reading sources... [ 35%] generated/anndata.AnnData.var_names
reading sources... [ 36%] generated/anndata.AnnData.var_names_make_unique
reading sources... [ 37%] generated/anndata.AnnData.var_vector
reading sources... [ 38%] generated/anndata.AnnData.varm
reading sources... [ 39%] generated/anndata.AnnData.varm_keys
reading sources... [ 39%] generated/anndata.AnnData.varp
reading sources... [ 40%] generated/anndata.AnnData.write
reading sources... [ 41%] generated/anndata.AnnData.write_csvs
reading sources... [ 42%] generated/anndata.AnnData.write_h5ad
reading sources... [ 43%] generated/anndata.AnnData.write_loom
reading sources... [ 44%] generated/anndata.AnnData.write_zarr
reading sources... [ 45%] generated/anndata.ImplicitModificationWarning
reading sources... [ 46%] generated/anndata.abc.CSCDataset
reading sources... [ 46%] generated/anndata.abc.CSCDataset.__getitem__
reading sources... [ 47%] generated/anndata.abc.CSCDataset.backend


Versions
========

* Platform:         darwin; (macOS-15.3.1-arm64-arm-64bit)
* Python version:   3.12.5 (CPython)
* Sphinx version:   8.2.1
* Docutils version: 0.21.2
* Jinja2 version:   3.1.5
* Pygments version: 2.19.1

Last Messages
=============

    generated/anndata.abc.CSCDataset


    reading sources... [ 46%]
    generated/anndata.abc.CSCDataset.__getitem__


    reading sources... [ 47%]
    generated/anndata.abc.CSCDataset.backend

Loaded Extensions
=================

* sphinx.ext.mathjax (8.2.1)
* alabaster (1.0.0)
* sphinxcontrib.applehelp (2.0.0)
* sphinxcontrib.devhelp (2.0.0)
* sphinxcontrib.htmlhelp (2.1.0)
* sphinxcontrib.serializinghtml (2.0.0)
* sphinxcontrib.qthelp (2.0.0)
* myst_parser (4.0.1)
* sphinx_copybutton (0.5.2)
* sphinx.ext.autodoc.preserve_defaults (8.2.1)
* sphinx.ext.autodoc.type_comment (8.2.1)
* sphinx.ext.autodoc.typehints (8.2.1)
* sphinx.ext.autodoc (8.2.1)
* sphinx.ext.intersphinx (8.2.1)
* sphinx.ext.doctest (8.2.1)
* sphinx.ext.coverage (8.2.1)
* sphinx.ext.napoleon (8.2.1)
* sphinx.ext.autosummary (8.2.1)
* sphinx_autodoc_typehints (unknown version)
* sphinx_issues (5.0.0)
* sphinx_design (0.6.1)
* sphinx_search.extension (0.3.2)
* sphinxext.opengraph (unknown version)
* scanpydoc.definition_list_typed_field (0.15.2)
* scanpydoc.elegant_typehints (0.15.2)
* scanpydoc.rtd_github_links (0.15.2)
* scanpydoc.theme (unknown version)
* scanpydoc.release_notes (0.15.2)
* scanpydoc (0.15.2)
* sphinx.ext.linkcode (8.2.1)
* nbsphinx (0.9.6)
* IPython.sphinxext.ipython_console_highlighting (unknown version)
* sphinx_toolbox.more_autodoc.autoprotocol (3.8.3)
* no_skip_abc_members (unknown version)
* patch_myst_cite (unknown version)
* sphinx_book_theme (unknown version)
* pydata_sphinx_theme (unknown version)

Traceback
=========

      File "/Users/srivarra/Library/Application Support/hatch/env/virtual/anndata/_5ujvw4a/docs/lib/python3.12/site-packages/sphinx/ext/autodoc/__init__.py", line 2980, in add_directive_header
        objrepr = stringify_annotation(
                  ^^^^^^^^^^^^^^^^^^^^^
    TypeError: _stringify_annotation() got an unexpected keyword argument 'short_literals'


The full traceback has been saved in:
/var/folders/fy/q2szypn9325d_0g0nq049k300000gq/T/sphinx-err-k20ijzbv.log

To report this error to the developers, please open an issue at <https://github.com/sphinx-doc/sphinx/issues/>. Thanks!
Please also report this if it was a user error, so that a better error message can be provided next time.
```

</details>

I made sure to merge `main` into this branch, but idk.

In addition I can't quite comprehend this single test which didn't pass [`MEMORY PROBLEMS tests/test_dask_view_mem.py::test_size_of_view[layers-True]
`][azure-dev-test-fail]. It did pass locally on my mac however.

[azure-dev-test-fail]:https://dev.azure.com/scverse/anndata/_build/results?buildId=9555&view=logs&j=ed19f947-f4ce-5e43-250d-4ce4db49a756&t=de40599c-6b0c-5ef7-3ea6-59cb0efad3c6&l=569 

Thank you!

ilan-gold commented:
> I'm not too sure what's going on with the readthedocs build, since I'm unable to replicate it locally, instead I have a different error:

Remotely, the issue is that you're exporting a new function whose typing depends on `anndata._core.extensions.ExtensionNamespace` so you'll need to export that as well in the docs.

If you're running the docs using `hatch` you can prune your environment - if not, would highly recommend using hatch! Check out our `hatch.toml`

> In addition I can't quite comprehend this single test which didn't pass [MEMORY PROBLEMS tests/test_dask_view_mem.py::test_size_of_view[layers-True] ](https://dev.azure.com/scverse/anndata/_build/results?buildId=9555&view=logs&j=ed19f947-f4ce-5e43-250d-4ce4db49a756&t=de40599c-6b0c-5ef7-3ea6-59cb0efad3c6&l=569). It did pass locally on my mac however.

That one is just a hair flaky, we see this every so often, but generally passes (as it does now)

ilan-gold commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,269 @@
+
+
+@runtime_checkable
+class ExtensionNamespace(Protocol[NS]):
```
I don't see this protocol used anywhere.  It should be used in most if not all occurrences of `type[NS]` unless I am mistaken.

srivarra replied:
@ilan-gold
 
> I don't see this protocol used anywhere. It should be used in most if not all occurrences of type[NS] unless I am mistaken.

Ah I see, I was learning about them for the changes, and wasn't really sure how they are used, that make sense. Looks like it's how the other protocols for read/write are used.

----

> If each transform is intended only to do one thing (like SinhTransform only implements adata.transforms.sinh) then maybe a _transform function annotated properly should be implemented and then used by the public-facing API like adata.transform.sinh i.e., it calls _run to do the actual work. But this wouldn't help with the public-facing API at least.

As I understand it, the suggestion is that each transformation method should have its own class right? Here are my thoughts on that approach.

I think each namespace should be able to do multiple things, as accessors / extensions are a grouping of specific subdomain functionality. The namespace is a container for those that extended functionality. Instead of multiple classes which implement different transformations, one transformation class should be able to implement multiple methods.

When considering the Transform example, between: 
```python
@ad.register_anndata_namespace("sinh")
class SinhTransform:
    """Accessor for applying inverse hyperbolic sine transformation."""
    
    def __init__(self, adata: ad.AnnData):
        self._adata = adata
    
    def transform(
        self, 
        layer: Optional[str] = None, 
        scale: float = 1.0,
        output_layer: Optional[str] = None,
        inplace: bool = False
    ) -> Optional[ad.AnnData]: ...


@ad.register_anndata_namespace("log1p")
class Log1pTransform:
    """Accessor for applying log1p transformation."""
    
    def __init__(self, adata: ad.AnnData):
        self._adata = adata
    
    def transform(
        self, 
        layer: Optional[str] = None, 
        scale: float = 1.0,
        output_layer: Optional[str] = None,
        inplace: bool = False
    ) -> Optional[ad.AnnData]: ...

```

and 

```python
import anndata as ad
import numpy as np
from typing import Optional, Union, Callable, Literal


@ad.register_anndata_namespace("transform")
class TransformNamespace:
    """Accessor for applying various transformations to AnnData objects."""

    def __init__(self, adata: ad.AnnData):
        self._adata = adata

    def sinh(
        self,
        layer: Optional[str] = None,
        scale: float = 1.0,
        output_layer: Optional[str] = None,
        inplace: bool = False,
    ) -> Optional[ad.AnnData]: ...

    def log1p(
        self,
        layer: Optional[str] = None,
        base: Optional[float] = None,
        output_layer: Optional[str] = None,
        inplace: bool = False,
    ) -> Optional[ad.AnnData]: ...

```

I prefer the 2nd option due to less namespace collision making it easier for the user and for the developer of extensions to handle. For example in one of my packages I put all the related functionality in one public facing class [here][annsel-example].

-----

> In general, it would be nice to have some sort of guarantee on the function that the subclass of ExtensionNamespace implements i.e., sinh or log1p.

> How does polars handle this? Does each transform only do one thing or can one do multiple? I'd be tempted to make sinh for example, a string that is then used to look up an attribute etc. to call the function.

For these two, there isn't a set in stone list of guidelines that I've seen, but I can give an overview of what others have implemented. Polar's doesn't set guarantees.

The extension library [Polars OLS][polars-ols-link] writes their functions s.t. an [extension on an `Expr` also returns an `Expr` type][polars-ols-register].

In the official docs the examples they give mostly follow the same structure that the extension on the type returns that type (i.e. `Expr` extensions return `Expr`, `pl.DataFrames` return `pl.DataFrames`). There was one odd one out which is the [df][polars-docs-api-df] example where a function returns a list of `pl.DataFrames`.

Polars doesn't impose strict rules on how users should structure their extensions. In practice, most extension implementations follow the principle of having each method perform a single, well-defined function - adhering to the "do one thing well" philosophy. This approach is also common in other libraries with extension systems, such as xarray's accessor pattern [xarray-accessors-tutorial].

But I think a person making the extension should be able to return whatever they want with those functions. For example here is a plotting extension which has `heatmap` method which returns a `matplotlib` `Figure` object.
<details>
<summary>Plotting accessor example</summary>

```python
import anndata as ad
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from typing import Optional, Tuple


@ad.register_anndata_namespace("plot")
class PlottingAccessor:
    def __init__(self, adata: ad.AnnData):
        self._adata = adata

    def heatmap(
        self,
        layer: Optional[str] = None,
        vars: Optional[list[str]] = None,
        obs_subset: Optional[list[str]] = None,
        cmap: str = "viridis",
        figsize: Tuple[float, float] = (10, 8),
        title: Optional[str] = None,
        **kwargs,
    ) -> plt.Figure:
        """
        Plot a heatmap of gene expression.

        Parameters
        ----------
        layer
            Layer to use for plotting. If None, uses X.
        vars
            List of var names to include. If None, uses all vars.
        obs_subset
            List of observation names to include. If None, uses all observations.
        cmap
            Colormap to use for the heatmap.
        figsize
            Figure size as (width, height) in inches.
        title
            Title for the plot. If None, a default title is generated.
        **kwargs
            Additional keyword arguments passed to seaborn.heatmap.

        Returns
        -------
        Figure object
        """
        # Get the data to plot
        if layer is not None:
            if layer not in self._adata.layers:
                raise ValueError(f"Layer {layer} not found in AnnData object")
            data = self._adata.layers[layer]
        else:
            data = self._adata.X

        # Convert to dense array if sparse
        if hasattr(data, "toarray"):
            data = data.toarray()

        # Subset genes if specified
        if vars is not None:
            if not all(var in self._adata.var_names for var in vars):
                missing = [v for v in vars if v not in self._adata.var_names]
                raise ValueError(f"Variables not found in AnnData object: {missing}")
            var_idx = [self._adata.var_names.get_loc(var) for var in vars]
            data = data[:, var_idx]
            var_names = vars
        else:
            var_names = self._adata.var_names

        # Subset observations if specified
        if obs_subset is not None:
            if not all(obs in self._adata.obs_names for obs in obs_subset):
                missing = [o for o in obs_subset if o not in self._adata.obs_names]
                raise ValueError(f"Observations not found in AnnData object: {missing}")
            obs_idx = [self._adata.obs_names.get_loc(obs) for obs in obs_subset]
            data = data[obs_idx, :]
            obs_names = obs_subset
        else:
            obs_names = self._adata.obs_names

        # Create the plot
        fig, ax = plt.subplots(figsize=figsize)
        sns.heatmap(
            data,
            cmap=cmap,
            xticklabels=var_names if len(var_names) < 50 else False,
            yticklabels=obs_names if len(obs_names) < 50 else False,
            ax=ax,
            **kwargs,
        )

        # Set title
        if title is None:
            title = f"Expression Heatmap{f' ({layer})' if layer else ''}"
        ax.set_title(title)

        plt.tight_layout()
        return fig

## Used like so:

rng = np.random.default_rng(42)
n_obs, n_vars = 100, 50
X = rng.normal(0, 1, size=(n_obs, n_vars))

# Create some metadata
import pandas as pd

obs = pd.DataFrame(
    {
        "cell_type": rng.choice(["T cell", "B cell", "NK cell"], size=n_obs),
        "condition": rng.choice(["control", "treatment"], size=n_obs),
    }
)

var = pd.DataFrame(
    {"gene_type": rng.choice(["protein_coding", "lncRNA"], size=n_vars)},
    index=[f"gene_{i}" for i in range(n_vars)],
)

# Create the AnnData object
adata = ad.AnnData(X=X, obs=obs, var=var)

# Use the plotting accessor
fig = adata.plot.heatmap(
    vars=["gene_0", "gene_1", "gene_2"], title="Gene Expression Heatmap"
)

```

</details>


[polars-ols-link]: https://github.com/azmyrajab/polars_ols

[polars-ols-register]: https://github.com/azmyrajab/polars_ols/blob/119f98475c32575b3a54b34caccb63deae4e4f81/polars_ols/__init__.py#L36

[polars-docs-api-df]: https://docs.pola.rs/api/python/stable/reference/api.html

[xarray-accessors-tutorial]: https://tutorial.xarray.dev/advanced/accessors/01_accessor_examples.html

[annsel-example]: https://github.com/srivarra/annsel/blob/28b441d8fd3b54fdecc650367b9e710b3b640191/src/annsel/tl/anndata.py#L23

ilan-gold replied:
> Ah I see, I was learning about them for the changes, and wasn't really sure how they are used, that make sense. Looks like it's how the other protocols for read/write are used.

No worries, this PR is super great and easy to review, great job!

> As I understand it, the suggestion is that each transformation method should have its own class right?

I guess that's what I'm asking.  Is that the intention?

> But I think a person making the extension should be able to return whatever they want with those functions. For example here is a plotting extension which has heatmap method which returns a matplotlib Figure object.

Ah interesting, true, that would be good.  I hadn't considered that use-case.  Perhaps then the `__init__` is enough.  I was thinking too narrowly.

ilan-gold replied:
I am fine with the second one, looks good to me

ilan-gold commented on .vscode/settings.json:
```diff
@@ -15,7 +15,7 @@
         "editor.defaultFormatter": "biomejs.biome",
     },
-    "python.analysis.typeCheckingMode": "basic",
+    "python.analysis.typeCheckingMode": "off",
```
Please undo

ilan-gold commented on src/anndata/__init__.py:
```diff
@@ -68,11 +69,13 @@ def __getattr__(attr_name: str) -> Any:
     # Classes
     "AnnData",
     "Raw",
+    "ExtensionNamespace",
```
We have a `typing.py` file so you don't have to reexport

srivarra replied:
Forgot to remove it from there. Should the protocol go in `src/anndata/_types.py` or `src/anndata/typing.py`?

ilan-gold replied:
@flying-sheep can answer this best, but I think `_types.py` actually

flying-sheep replied:
The names have the following meaning `typing` is for type checking, `types` have runtime checkable things like protocols with `@runtime_checkable` and so on.

And nothing with `_leading_underscore` should ever appear in docs.

ilan-gold commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,269 @@
+    try:
+        type_hints = get_type_hints(ns_class.__init__)
+        resolved_type = type_hints.get(param.name, param.annotation)
+    except Exception:
```
Please remove the blanket `except` here with something more precise

srivarra commented on src/anndata/_types.py:
```diff
@@ -186,3 +188,23 @@ def __call__(
+
+
+@runtime_checkable
+class ExtensionNamespace(Protocol[NS]):
```
Just thought it was necessary with the protocol, but looks like it doesn't really do anything.

flying-sheep commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,271 @@
+
+def register_anndata_namespace(
+    name: str,
+) -> Callable[[type[ExtensionNamespace]], type[ExtensionNamespace]]:
```
and here

flying-sheep commented on src/anndata/_core/extensions.py:
```diff
@@ -0,0 +1,271 @@
+
+    Returns
+    -------
+        A decorator that registers the decorated class as a custom namespace.
```
```suggestion
    A decorator that registers the decorated class as a custom namespace.
```

ilan-gold commented on docs/api.md:
```diff
@@ -158,6 +158,25 @@ Types used by the former:
+.. autosummary::
+   :toctree: generated/
+
+   _types.ExtensionNamespace
```
what would the proposal here be then? make a public `types` package?

flying-sheep replied:
… or the other thing I mentioned:

> put it somewhere public or don’t refer to its name at all in the docs

ilan-gold commented on tests/test_extensions.py:
```diff
@@ -0,0 +1,241 @@
+                    self._adata = adata
+
+
+class TestNamespaceSignatureValidation:
```
What are you referring to here?

flying-sheep replied:
ah my bad. that’s what I get for commenting when sick.

I thought this called `register_anndata_namespace` at module execution time, but it only calls it inside of test functions as it should.

flying-sheep reviewed (commented):
Apart from the above concern about having `_private` names in the docs, this looks nice!

I fixed the `__get__` typing and reused the fixture in the tests some more.
